File: picasso-transformations/src/main/java/io/github/hendraanggrian/picassotransformations/GrayscaleTransformation.java
package io.github.hendraanggrian.picassotransformations;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;

import com.squareup.picasso.Transformation;

import io.github.hendraanggrian.picassotransformations.internal.PaintBuilder;

/**
 * @author Hendra Anggrian (hendraanggrian@gmail.com)
 */
class GrayscaleTransformation implements Transformation {

    static final String TAG = "GrayscaleTransformation";

    @Override
    public Bitmap transform(Bitmap source) {
        final Bitmap bitmap = Bitmap.createBitmap(source.getWidth(), source.getHeight(), Bitmap.Config.ARGB_8888);
        final ColorMatrix saturation = new ColorMatrix();
        saturation.setSaturation(0f);
        new Canvas(bitmap).drawBitmap(source, 0, 0, new PaintBuilder()
                .colorFilter(new ColorMatrixColorFilter(saturation))
                .build());
        source.recycle();
        return bitmap;
    }

    @Override
    public String key() {
        return TAG;
    }
}

File: pikasso-transformations/src/com/hendraanggrian/pikasso/transformations/GrayscaleTransformation.kt
package com.hendraanggrian.pikasso.transformations

import android.graphics.Bitmap
import android.graphics.Bitmap.Config.ARGB_8888
import android.graphics.Bitmap.createBitmap
import android.graphics.Canvas
import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.graphics.Paint
import com.squareup.picasso.Transformation

@PublishedApi
internal class GrayscaleTransformation : Transformation {

    override fun transform(source: Bitmap): Bitmap {
        val target = createBitmap(source.width, source.height, ARGB_8888)
        Canvas(target).drawBitmap(source, 0f, 0f, Paint().apply {
            colorFilter = ColorMatrixColorFilter(ColorMatrix().apply { setSaturation(0f) })
        })
        source.recycle()
        return target
    }

    override fun key(): String = "GrayscaleTransformation()"
}
